plenum/server/consensus/ordering_service.py

```diff
@@ -2143,3 +2153,298 @@ def generateName(node_name: str, inst_id: int):
+        self.last_accepted_pre_prepare_time = tm
+        if self.is_master and rejects:
+            self.send_outbox(rejects)
+        return pre_prepare
+
+    def l_get_last_timestamp_from_state(self, ledger_id):
+        if ledger_id == DOMAIN_LEDGER_ID:
+            ts_store = self.db_manager.get_store(TS_LABEL)
+            if ts_store:
+                last_timestamp = ts_store.get_last_key()
+                if last_timestamp:
+                    last_timestamp = int(last_timestamp.decode())
+                    self._logger.debug("Last ordered timestamp from store is : {}"
+                                      "".format(last_timestamp))
+                    return last_timestamp
+        return None
+
+    # This is to enable replaying, inst_id, view_no and pp_seq_no are used
+    # while replaying
+    def l_get_utc_epoch_for_preprepare(self, inst_id, view_no, pp_seq_no):
+        tm = self.get_time_for_3pc_batch()
+        if self.last_accepted_pre_prepare_time and \
+                tm < self.last_accepted_pre_prepare_time:
+            tm = self.last_accepted_pre_prepare_time
+        return tm
+
+    def l_consume_req_queue_for_pre_prepare(self, ledger_id, tm,
+                                          view_no, pp_seq_no):
+        reqs = []
+        rejects = []
+        invalid_indices = []
+        idx = 0
+        while len(reqs) < self._config.Max3PCBatchSize \
+                and self.requestQueues[ledger_id]:
+            key = self.requestQueues[ledger_id].pop(0)
+            if key in self._requests:
+                fin_req = self._requests[key].finalised
+                malicious_req = False
+                try:
+                    self.l_processReqDuringBatch(fin_req,
+                                               tm)
+
+                except (
+                        InvalidClientMessageException,
+                        UnknownIdentifier
+                ) as ex:
+                    self._logger.warning('{} encountered exception {} while processing {}, '
+                                        'will reject'.format(self, ex, fin_req))
+                    rejects.append((fin_req.key, Reject(fin_req.identifier, fin_req.reqId, ex)))
+                    invalid_indices.append(idx)
+                except SuspiciousPrePrepare:
+                    malicious_req = True
+                finally:
+                    if not malicious_req:
+                        reqs.append(fin_req)
+                if not malicious_req:
+                    idx += 1
+            else:
+                self._logger.debug('{} found {} in its request queue but the '
+                                  'corresponding request was removed'.format(self, key))
+
+        return reqs, invalid_indices, rejects
+
+    def l_sendPrePrepare(self, ppReq: PrePrepare):
+        self.sentPrePrepares[ppReq.viewNo, ppReq.ppSeqNo] = ppReq
+        self.l_send(ppReq, TPCStat.PrePrepareSent)
+
+    def l_send(self, msg, stat=None) -> None:
```

Comment:
Should we use this method in all places (in OrderingService) where we call `self._network.send`?